File: LeetCode/530_Minimum_Absolute_Difference_in_BST/Solution.java
import java.util.List;
import java.util.ArrayList;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Solution {
    public int getMinimumDifference(TreeNode root) {
        List<Integer> numbers = new ArrayList<>();

        inOrderTraversal(root, numbers);

        int min = Integer.MAX_VALUE;
        for (int i = 1; i < numbers.size(); i++) {
            min = Math.min(numbers.get(i) - numbers.get(i - 1), min);
        }
        return min;
    }

    public void inOrderTraversal(TreeNode root, List<Integer> numbers) {
        if (root == null)
            return;

        inOrderTraversal(root.left, numbers);
        numbers.add(root.val);
        inOrderTraversal(root.right, numbers);
    }
}

File: LeetCode/530_Minimum_Absolute_Difference_in_BST/Solution.c
#include <limits.h>
#include <stdlib.h>

#define MIN(a, b) ((a) < (b) ? (a) : (b))

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

int count_nodes(struct TreeNode *node)
{
    if (node == NULL) {
        return 0;
    }

    int left = count_nodes(node->left);
    int right = count_nodes(node->right);
    return left + right + 1;
}

void inorder_traversal(struct TreeNode *node, int *numbers, int *index)
{
    if (node == NULL)
        return;

    inorder_traversal(node->left, numbers, index);
    numbers[*index] = node->val;
    (*index)++;
    inorder_traversal(node->right, numbers, index);
}

int getMinimumDifference(struct TreeNode *root)
{
    int n = count_nodes(root);
    int *numbers = calloc(n, sizeof(int));
    int result = INT_MAX;
    int index = 0;

    inorder_traversal(root, numbers, &index);

    for (int i = 1; i < n; i++) {
        result = MIN(numbers[i] - numbers[i - 1], result);
    }
    return result;
}
